components/feature/prompts/src/main/java/mozilla/components/feature/prompts/PromptFeature.kt

```diff
@@ -126,7 +137,23 @@ class PromptFeature(
      */
     override fun start() {
         promptAbuserDetector.resetJSAlertAbuseState()
-        observer.observeIdOrSelected(sessionId)
+
+        scope = store.flowScoped { flow ->
+            flow.map { state -> state.findCustomTabOrSelectedTab(customTabId) }
+                .ifAnyChanged {
+                    arrayOf(it?.content?.promptRequest, it?.content?.loading)
+                }
+                .collect { state ->
+                    state?.content?.let {
+                        if (it.promptRequest != activePromptRequest) {
+                            onPromptRequested(state)
+                        } else if (!it.loading) {
+                            promptAbuserDetector.resetJSAlertAbuseState()
+                        }
+                        activePromptRequest = it.promptRequest
+                    }
```

Comment:
The `ifAnyChanged()` is fancy. But it makes the collect more complicated to digest because you now need to `if else` check what actually changed. It doesn't need to block landing this - but we should consider thinking about alternatives eventually.

Reply:
Yes, agreed! I think we need API that allows handling individual changes, but this would be a bigger change. Hope we can work on that later this month!